README.md
# r-koopa

Internal R package dependency for [koopa](https://koopa.acidgenomics.com/).

```r
if (!requireNamespace("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
}
install.packages(
    pkgs = "koopa",
    repos = c(
        "https://r.acidgenomics.com",
        BiocManager::repositories()
    )
)
```


NEWS.md
## koopa 0.1.20 (2021-07-06)

### Minor changes

- Add system checks for pyenv and poetry.

## koopa 0.1.19 (2021-06-07)

### Minor changes

- Updated dependency version cutoffs.

## koopa 0.1.18 (2021-05-11)

### Minor changes

- Added back version-specific checks for AWS CLI, Azure CLI, and
  Google Cloud SDK.

## koopa 0.1.17 (2021-05-09)

### Minor changes

- Updated system checks and improved installation instructions.

## koopa 0.1.16 (2021-04-30)

### Minor changes

- `checkSystem`: Removed check for deprecated `pandoc-citeproc` package.
- `checkSystem`: Ensure that conda is available in the PATH for version check.

## koopa 0.1.15 (2021-04-23)

### New functions

- `gitDefaultBranch`: Detect the default branch of a git repository (e.g. "main"
  or "master").

### Minor changes

- `drat`: Update to avoid hard-coding against default "main" or "master" branch.

## koopa 0.1.14 (2021-04-09)

### Minor changes

- Update system checks to reflect more minimal GNU config on macOS.

## koopa 0.1.13 (2021-04-06)

### Minor changes

- `checkSystem`: Add fzf and hadolint system checks.

## koopa 0.1.12 (2021-03-30)

### Minor changes

- `dockerPruneStaleTags`: Updated to use environment variables for DockerHub
  credentials.

## koopa 0.1.11 (2021-03-18)

### Minor changes

- `drat`: Changed archive mode to play nicer with AWS CloudFront.
- `checkSystem`: Updated Ruby checks.

## koopa 0.1.10 (2021-03-10)

### New functions

- Added `dockerPruneTags`, which helps prune old tags automatically.
- Bug fix for prefix handling in syntactic rename functions.

## koopa 0.1.9 (2021-03-02)

### Minor changes

- Added more package version checks: boost, cairo, harfbuzz, icu4c, imagemagick,
  and node.

## koopa 0.1.8 (2021-03-01)

### Minor changes

- Added file size check to `drat`.
- Updated dependency packages.
- Cleaned up NAMESPACE.

## koopa 0.1.7 (2021-02-17)

- `drat`: pkgdown documentation in docs is no longer removed.

## koopa 0.1.6 (2021-02-16)

- `dockerBuildAllTags`: Ignore lines that contain comments in `build.txt` files.

## koopa 0.1.5 (2021-02-15)

- Migrated `checkBinManConsistency` here from main koopa package.

## koopa 0.1.4 (2021-02-15)

- Added version check for GNU tar.
- Updated dependency version cutoffs.

## koopa 0.1.3 (2021-02-11)

- Including checking for GNU binutils on macOS in addition to Linux.

## koopa 0.1.2 (2021-02-06)

- Improved internal argument parsing for syntactic functions.
- `drat`: Added optional override flags: `--no-check`, `--no-pkgdown`,
  and `--no-deploy`.

## koopa 0.1.1 (2021-02-04)

- Updated internal bb8 dependency to renamed AcidDevTools package.
- Migrated `detectHPC` function from basejump, which is more useful for koopa.

## koopa 0.1.0 (2021-01-31)

- Initial stable public release.
- Bug fixes for `drat`.
- Improved genome download support for Ensembl, Gencode, RefSeq, and UCSC.

## koopa 0.0.23 (2020-01-04)

- Added check for bpytop.
- Added check for GNU stow.
- `listPrograms` now excludes `app` directory, as expected.
- Migrated internal R scripts previously defined in koopa package.

## koopa 0.0.22 (2020-12-24)

- Bug fix for GPG Suite version check on macOS.

## koopa 0.0.21 (2020-12-14)

- `parseArgs`: Bug fix for parsing of quoted arguments containing spaces.

## koopa 0.0.20 (2020-12-07)

- Added `hasPositionalArgs` boolean check for user input.

## koopa 0.0.19 (2020-11-19)

- Added groff to system check.
- Removed autojump from system check. Now favoring zoxide instead.

## koopa 0.0.18 (2020-11-12)

- Internal syntax updates reflecting reorganization in main `koopa` program.

## koopa 0.0.17 (2020-11-10)

- `koopaPrefix`: Path resolution fix on Linux.
  Needed to wrap `koopa` bin path in `dirname` call.

## koopa 0.0.16 (2020-11-10)

- Added checks for Rust dog, procs, and starship.
- Added check for ps as a base dependency.

## koopa 0.0.15 (2020-11-06)

- Now requiring bb8 0.2.41+.

## koopa 0.0.14 (2020-11-04)

- Added more suggested packages: BiocManager, covr, desc, lintr, remotes, and
  sessioninfo.
- Removed defunct `installBioconductor` reexport.

## koopa 0.0.13 (2020-11-03)

- Migrated installer functions to bb8: `installAcidverse`,
  `installBioconductor`, and `installDefaultPackages`.

## koopa 0.0.12 (2020-10-29)

- `installDefaultPackages`: Added temporary fix for DelayedArray not currently
  available for Bioc Devel (3.13).
  Requires goalie v0.4.10, which adds `isBiocDevel` check function.
  Also requires bb8 v0.2.34, which adds easy installation from Git repos
  inside the `install` function.

## koopa 0.0.11 (2020-10-28)

- Added new `installAcidverse` function.
- Updated `installDefaultPackages` to call `installAcidverse` internally.

## koopa 0.0.10 (2020-10-26)

- Added support for checking armadillo version.

## koopa 0.0.9 (2020-10-06)

- Renamed acidbase dependency to AcidBase.

## koopa 0.0.8 (2020-09-11)

- Added `realpath` to reexports.

## koopa 0.0.7 (2020-09-08)

- `installDefaultPackages`: Added broom, biobroom, and some other recommended
  packages from r-lib.

## koopa 0.0.6 (2020-08-27)

- `installDefaultPackages`: Removed temporary mgcv build fix for macOS.

## koopa 0.0.5 (2020-08-25)

- `dockerBuildAllTags`: Fix for recently built image check.

## koopa 0.0.4 (2020-08-25)

- `installDefaultPackages`: Added PANTHER.db package.

## koopa 0.0.3 (2020-08-24)

- `installDefaultPackages`: Bug fix for mgcv compilation issue on macOS.

## koopa 0.0.2 (2020-08-17)

- Updated `installDefaultPackages` to install from Acid Genomics drat repo
  rather than from GitHub directly.

## koopa 0.0.1 (2020-08-12)

- Initial commit of koopa R package.


R/drat.R
#' Build a package and commit to drat repo
#'
#' @name drat
#' @note Updated 2021-04-23.
#'
#' @return `logical(1)`.
#'
#' @examples
#' ## > drat()
NULL



.drat <- function(
    pkgDirs,
    repoDir = file.path("~", "monorepo", "drat"),
    check = TRUE,
    pkgdown = TRUE,
    deploy = TRUE
) {
    requireNamespaces(c("desc", "devtools"))
    assert(
        allAreDirs(pkgDirs),
        isADir(repoDir),
        isFlag(check),
        isFlag(pkgdown),
        isFlag(deploy)
    )
    pkgDirs <- realpath(pkgDirs)
    if (isTRUE(deploy)) {
        wd <- getwd()
    }
    lapply(
        X = pkgDirs,
        repoDir = repoDir,
        check = check,
        pkgdown = pkgdown,
        deploy = deploy,
        FUN = function(pkgDir, repoDir, check, pkgdown, deploy) {
            descFile <- file.path(pkgDir, "DESCRIPTION")
            assert(isAFile(descFile))
            if (isTRUE(check)) {
                check(path = pkgDir)
            }
            if (isTRUE(pkgdown)) {
                .pkgdownDeployToAWS(package = pkgDir)
            }
            tarball <- devtools::build(
                pkg = pkgDir,
                vignettes = check
            )
            assert(isAFile(tarball))
            size <- file.size(tarball)
            if (isTRUE(size > 2e6L)) {
                stop(sprintf(
                    "Package tarball is too large: '%s'.",
                    tarball
                ))
            }
            drat::insertPackage(
                file = tarball,
                repodir = repoDir,
                ## NOTE "archive" is preferred but doesn't currently work
                ## perfectly with AWS CloudFront enabled.
                action = "none"
            )
            invisible(file.remove(tarball))
            if (isTRUE(deploy)) {
                setwd(pkgDir)
                name <- desc::desc_get_field(key = "Package", file = descFile)
                version <- as.character(desc::desc_get_version(file = descFile))
                today <- Sys.Date()
                shell(command = "git", args = c("fetch", "--force", "--tags"))
                shell(
                    command = "git",
                    args = c(
                        "tag",
                        "--force",
                        "-a",
                        paste0("'v", version, "'"),
                        "-m",
                        paste0("'", name, " v", version, " (", today, ")'")
                    )
                )
                shell(command = "git", args = c("push", "--force", "--tags"))
                setwd(repoDir)
                branch <- gitDefaultBranch(repoDir)
                shell(command = "git", args = c("checkout", branch))
                shell(command = "git", args = c("fetch", "--all"))
                shell(command = "git", args = "merge")
                shell(command = "git", args = c("add", "./"))
                shell(
                    command = "git",
                    args = c(
                        "commit", "-m", paste0("'Add ", basename(tarball), ".'")
                    )
                )
                shell(command = "git", args = "push")
                setwd(wd)
            }
            message(sprintf("Successfully added '%s'.", basename(tarball)))
            TRUE
        }
    )
    if (isTRUE(deploy)) {
        setwd(repoDir)
        assert(isAFile("update"))
        shell(command = "./update")
        setwd(wd)
    }
    invisible(TRUE)
}



#' @rdname drat
#' @export
drat <- function() {
    parse <- parseArgs(
        optional = "repo-dir",
        flags = c(
            "fast",
            "no-check",
            "no-deploy",
            "no-pkgdown"
        ),
        positional = TRUE
    )
    optional <- parse[["optional"]]
    flags <- parse[["flags"]]
    positional <- parse[["positional"]]
    args <- list(
        "pkgDirs" = positional,
        "check" = !isSubset("no-check", flags),
        "deploy" = !isSubset("no-deploy", flags),
        "pkgdown" = !isSubset("no-pkgdown", flags)
    )
    if (isSubset("fast", flags)) {
        args[["check"]] <- FALSE
        args[["pkgdown"]] <- FALSE
    }
    if (!is.null(optional)) {
        if (isSubset("repo-dir", names(optional))) {
            args[["repoDir"]] <- optional[["repo-dir"]]
        }
    }
    do.call(what = .drat, args = args)
}
